homeassistant/components/alarmdecoder/__init__.py

```diff
@@ -1,167 +1,85 @@
+    CONF_HOST,
+    CONF_PORT,
+    CONF_PROTOCOL,
+    EVENT_HOMEASSISTANT_STOP,
 )
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import dt as dt_util
 
-DEVICE_SERIAL_SCHEMA = vol.Schema(
-    {
-        vol.Required(CONF_DEVICE_TYPE): "serial",
-        vol.Optional(CONF_DEVICE_PATH, default=DEFAULT_DEVICE_PATH): cv.string,
-        vol.Optional(CONF_DEVICE_BAUD, default=DEFAULT_DEVICE_BAUD): cv.string,
-    }
+from .const import (
+    CONF_DEVICE_BAUD,
+    CONF_DEVICE_PATH,
+    DEFAULT_ZONE_OPTIONS,
+    DOMAIN,
+    OPTIONS_ZONES,
+    PROTOCOL_SERIAL,
+    PROTOCOL_SOCKET,
+    SIGNAL_OPTIONS_UPDATE,
+    SIGNAL_PANEL_MESSAGE,
+    SIGNAL_REL_MESSAGE,
+    SIGNAL_RFX_MESSAGE,
+    SIGNAL_ZONE_FAULT,
+    SIGNAL_ZONE_RESTORE,
 )
 
-DEVICE_USB_SCHEMA = vol.Schema({vol.Required(CONF_DEVICE_TYPE): "usb"})
-
-ZONE_SCHEMA = vol.Schema(
-    {
-        vol.Required(CONF_ZONE_NAME): cv.string,
-        vol.Optional(CONF_ZONE_TYPE, default=DEFAULT_ZONE_TYPE): vol.Any(
-            DEVICE_CLASSES_SCHEMA
-        ),
-        vol.Optional(CONF_ZONE_RFID): cv.string,
-        vol.Optional(CONF_ZONE_LOOP): vol.All(vol.Coerce(int), vol.Range(min=1, max=4)),
-        vol.Inclusive(
-            CONF_RELAY_ADDR,
-            "relaylocation",
-            "Relay address and channel must exist together",
-        ): cv.byte,
-        vol.Inclusive(
-            CONF_RELAY_CHAN,
-            "relaylocation",
-            "Relay address and channel must exist together",
-        ): cv.byte,
-    }
-)
+_LOGGER = logging.getLogger(__name__)
 
-CONFIG_SCHEMA = vol.Schema(
-    {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_DEVICE): vol.Any(
-                    DEVICE_SOCKET_SCHEMA, DEVICE_SERIAL_SCHEMA, DEVICE_USB_SCHEMA
-                ),
-                vol.Optional(
-                    CONF_PANEL_DISPLAY, default=DEFAULT_PANEL_DISPLAY
-                ): cv.boolean,
-                vol.Optional(CONF_AUTO_BYPASS, default=DEFAULT_AUTO_BYPASS): cv.boolean,
-                vol.Optional(
-                    CONF_CODE_ARM_REQUIRED, default=DEFAULT_CODE_ARM_REQUIRED
-                ): cv.boolean,
-                vol.Optional(CONF_ZONES): {vol.Coerce(int): ZONE_SCHEMA},
-            }
-        )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
+RESTART = False
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up for the AlarmDecoder devices."""
-    conf = config.get(DOMAIN)
+    return True
+
 
-    restart = False
-    device = conf[CONF_DEVICE]
-    display = conf[CONF_PANEL_DISPLAY]
-    auto_bypass = conf[CONF_AUTO_BYPASS]
-    code_arm_required = conf[CONF_CODE_ARM_REQUIRED]
-    zones = conf.get(CONF_ZONES)
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up AlarmDecoder config flow."""
+    entry.add_update_listener(_update_listener)
```

Comment:
`add_update_listener` returns a callable to cancel the listener.

The listener needs to be canceled on unload.

Reply:
done in https://github.com/home-assistant/core/pull/37998/commits/42a8cff068c03041b38c1fd31a2936203c3ab67c#diff-0084eacb87a2f8212f5bd9e29ca3282bR49 :+1: